Do not throw IOException on success

avdv opened the pull request:
Hi.

1. preparation
```console
$ touch /tmp/foo
$ sudo chown :1234 !$
```
2. run
```scala
import java.nio.file._
import java.nio.file.attribute

val p = Paths.get("/tmp/foo")

println(Files.readAttributes(p, classOf[attribute.PosixFileAttributes], LinkOption.NOFOLLOW_LINKS).group.getName)
```
throws an IOException (whereas there is no error actually):
```
java.io.IOException: Success
	at java.lang.Throwable.fillInStackTrace(Unknown Source)
	at java.lang.Throwable.init(Unknown Source)
	at java.lang.Exception.init(Unknown Source)
	at java.io.IOException.init(Unknown Source)
	at java.io.IOException.init(Unknown Source)
	at scala.scalanative.nio.fs.UnixException$.apply(Unknown Source)
	at scala.scalanative.nio.fs.NativePosixFileAttributeView.scala$scalanative$nio$fs$NativePosixFileAttributeView$$throwIOException(Unknown Source)
	at scala.scalanative.nio.fs.NativePosixFileAttributeView.scala$scalanative$nio$fs$NativePosixFileAttributeView$$getGroup(Unknown Source)
	at scala.scalanative.nio.fs.NativePosixFileAttributeView$$anon$2.scala$scalanative$nio$fs$NativePosixFileAttributeView$$anon$$fileGroup(Unknown Source)
	at scala.scalanative.nio.fs.NativePosixFileAttributeView$$anon$2$$anon$1$$anonfun$2.apply(Unknown Source)
	at scala.scalanative.nio.fs.NativePosixFileAttributeView$$anon$2$$anon$1$$anonfun$2.apply(Unknown Source)
	at scala.scalanative.native.Zone$.apply(Unknown Source)
	at scala.scalanative.nio.fs.NativePosixFileAttributeView$$anon$2$$anon$1.init(Unknown Source)
	at scala.scalanative.nio.fs.NativePosixFileAttributeView$$anon$2.group(Unknown Source)
	at Main$.main(Unknown Source)
```
Running the program on the JVM, simply prints out "1234" for the group name.

densh commented:
@LeeTibbert Can you please review this one?

LeeTibbert commented:
@avdv  Thank you for the PR and the opportunity to review it.  

I have spent some time reviewing it and the original code and studying
a few alternatives. You deftly moved through a house of cards!

I believe the code as submitted it correct. Every PR is a balance of
considerations.  Other things being equal, I value simpler changes,
but you are the person doing the work.

I would like to suggest a few simplifications which would
lead to fewer code changes.  It will take me until late tomorrow evening
to write them up for your consideration. 

I also believe that to the extent possible, every PR should have an accompanying 
test case.  The reproducer from this PR is a good start for the getGroup changes.
I have been focusing on alternate source changes, so I have not chased down if
an appropriate unit-test/*Suite.scala file exists or if it needs to be created.
I checked and at least on my machine creating a file and changing its owner
do not require sudo.  I seem to remember unit-tests under javalib creating 
and modifying files. They could be used as examples.    I can help with 
the fussy little details  after the source changes get settled if you like.

I know of no good way to unit-test the passwd changes (good catch there!),

Lee

LeeTibbert commented:
@avdv 

I sandboxed a number of alternatives and the following changes, based on your original,
seemed to be the simplest.  They worked for me (tests/test).  What do you think?
My goal is to honor your original work, not offend you.

They do not let nulls or synthesized data structures rattle around. The callers
of routines which can return missing values get to determine which action to take.

I left existing interfaces outside NativePosixFileAttributeView.scala alone. I dislike
the underlying interface but changing one part means changing a lot for consistency.
The changes reverberate and reinforce each sucking up time & life-force.

There is plenty of room for you or a scala expert to improve on these. 

1) No changes to .c files,  set errno to 0 in NativePosixFileAttributeView.scala
    & create another Issue that errno needs to be consistently set to 0 before
    each C call in any method that relies upon throwNewIOException. (long discussion here).

2) Change NativePosixFileAttributeView.scala  getGroup and getPasswd methods (4 in total
    to return an Option.
```
 private def getGroup(name: CString)(
      implicit z: Zone): Option[Ptr[grp.group]] = {
    val buf = alloc[grp.group]

    errno.errno = 0
    val err = grp.getgrnam(name, buf)

    if (err == 0) {
      Some(buf)
    } else if (errno.errno == 0) {
      None
    } else {
      throwIOException()
    }
  }

  private def getGroup(gid: stat.gid_t)(
      implicit z: Zone): Option[Ptr[grp.group]] = {
    val buf = alloc[grp.group]

    errno.errno = 0
    val err = grp.getgrgid(gid, buf)

    if (err == 0) {
      Some(buf)
    } else if (errno.errno == 0) {
      None
    } else {
      throwIOException()
    }
  }
private def getPasswd(name: CString)(
      implicit z: Zone): Option[Ptr[pwd.passwd]] = {
    val buf = alloc[pwd.passwd]

    errno.errno = 0
    val err = pwd.getpwnam(name, buf)

    if (err == 0) {
      Some(buf)
    } else if (errno.errno == 0) {
      None
    } else {
      throwIOException()
    }
  }

  private def getPasswd(uid: stat.uid_t)(
      implicit z: Zone): Option[Ptr[pwd.passwd]] = {
    val buf = alloc[pwd.passwd]

    errno.errno = 0
    val err = pwd.getpwuid(uid, buf)

    if (err == 0) {
      Some(buf)
    } else if (errno.errno == 0) {
      None
    } else {
      throwIOException()
    }
  }

```
3) Change the callers to handle the Option
```
override def setOwner(owner: UserPrincipal): Unit =
   Zone { implicit z =>
      val ownerName = owner.getName
      val passwd    = getPasswd(toCString(ownerName))

      if (passwd.isEmpty) {
        throw new UserPrincipalNotFoundException(ownerName)
      }

      errno.errno = 0
      if (unistd.chown(toCString(path.toString), passwd.get._2, -1.toUInt)
            != 0) {
        throwIOException()
      }
    }
 override def setGroup(group: GroupPrincipal): Unit =
    Zone { implicit z =>
      val grpName = group.getName
      val grp     = getGroup(toCString(grpName))

      if (grp.isEmpty) {
        throw new UserPrincipalNotFoundException(grpName)
      }

      errno.errno = 0
      if (unistd.chown(toCString(path.toString), -1.toUInt, grp.get._2) != 0) {
        throwIOException()
      }
    }
```

avdv commented:
Hi @LeeTibbert,

thank you for your reply, I very much appreciate it! And, I am in no way offended if someone posts an alternative solution to a problem, really!

The reason why I went with my approach was that I wanted to avoid "round-trips" between Scala and native code, that's why I pushed the reset of `errno` down to the C layer. Now I realize that it ends up as native code anyway, so there's not much difference (and the extra call probably will get optimized away at link time).

Having that said - I like your approach, let's go ahead and fix it. Would you take care of that? I assume you already have some code ready?

Claudio

densh commented:
@avdv Feel free to update your PR based on the feedback from @LeeTibbert. PR review eventually converges after a few rounds of changes and review.

avdv commented:
Hi, and sorry for the long delay.

I have incorporated the feedback, with only a few notably changes:

- use `Option.fold` instead of `Option.get`
- do not reset `errno` upfront of `chown` (`chown` clearly signals its error by the return value)

But, I also noticed a few other problems regarding the User / Group Principal classes.

First, the code assumes that you always are able to get the name for a user / group ID. This is not necessarily the case, just think about mounting an external file system with a different set of UIDs / GIDs.

Second, calling `Path.owner` or `Path.group` always issues a syscall (to determine the name). That's just wasting CPU cycles if you don't need the name.

Third, the User and Group Principal instances simply use object equality. That's a bit unfortunate when you want to group the files by user/group (but don't want to determine the name for each of your +10,000 files).

Note, I left the initial commits intact, in order to ease following along, instead of force pushing my changes. Not sure what you prefer...  I'll squash them once you want to merge.

BTW, I thought about splitting this into another issue, but since this simplifies the code in `NativePosixFileAttributeView` quite a bit I went with hijacking this one. :wink:

avdv commented on javalib/src/main/scala/scala/scalanative/nio/fs/NativePosixFileAttributeView.scala:
```diff
@@ -58,10 +64,13 @@ final class NativePosixFileAttributeView(path: Path, options: Array[LinkOption])
+      val gid = group match {
+        case g: NativeGroupPrincipal => g.gid
+
+        case _ => throw new IllegalArgumentException("unsupported GroupPrincipal")
```
We could also support "normal" principals, with only a name by doing a lookup here.

LeeTibbert replied:
Claudio,
Please let me test my understanding.  Looks to me like GroupPrinciple is at the
bottom of the type hierarchy and that NativeGroupPrinciple is the only
concrete implementer of GroupPrinciple.

What do you mean by "normal". Normal in the type hierarchy or normal as
meant by anyone actually trying to use the method?   Seems to me that the
code you quoted is correct for the former, and that the latter is the concern
of the caller of the method.   Seems like the routine in its JVM defined scope
is correct as is.

Am I off the rails here?  You have been exercising this area and know it.

avdv replied:
Hi Lee, 

yes, with "normal" I was meaning anyone else calling the method with some sort of custom `GroupPrincipal`.

I thought it should be safe to assume that a caller of the method got the GroupPrincipal by either calling the lookup method of the UserPrincipalLookupService or from other functions of the javalib module which all return a concrete `NativeGroupPrincipal`. (same goes for the UserPrincipal)

Sorry for the confusion!

LeeTibbert replied:
I thought about this a lot and wrote out a long complicated answer describing a
time consuming solution.  

Then it dawned on me that the code in question is part of java.nio and the only
people potentially going to be creating implementations of GroupPrincipal and UserPrincipal
are going to be SN devos.  

Seem to be that the code you quoted is a good sanity check for devos & corruption 
and that we do not need to restrict access to the Principal trait to thwart actively hostile users.

LeeTibbert commented:
Claudio,

It is invigorating to see this next evolution.  I have not been able to review line by line, but
every one of the lines I have looked at, and that is most, has been correct (no surprise to you).

I can do a detailed review Tuesday-ish, but do not want to stand in the way of this PR making
even more progress if something needs to happen before then.

I like the overall approach. It looks like it simplifies & improves the current state. Bravo!
I also noticed that you picked up the associated doc change for javalib.rst.

### Two high level responses:

* Do you know if existing unit-tests exercise any of this PR?  In particular, are
new unit-tests needed for the new to SN classes & their methods?
I can help write these if needed.

    While every bit of code should be correct, this PR deals with uid/gid issues and must
    be like Caesar's wife. My goal is to help improve the PR, not to be an impediment
    at mile 26  (kilometer ???) of a marathon.

 * The standing pain in the ... question that I ask myself with every
exception throw I write: If I can provoke this case on the JVM does the message I am
writing match?  If not, does it need to match or is there a SN reason for it not matching?
If the latter, I drop a one line comment in the code just before the throw.  Yeah, I 
am probably not scala-ish enough in that I like to leave at least a clue, not a volume for
the next poor bloke to read the code.  In a week, someone will probably be asking me
exactly the same question.

    As a reviewer, I like to ask the question in a supportive manner. Then again, much as I like JVM 
    matched messages, the perfect should not stand in the way of the acceptable & done. 
    The messages as written are clear and seem to provide the information available.

#### Point issues:
*  The use of `Option.fold` is elegant & appropriate. The best complement that I can
    give is that I would like to add that idiom to my bag of tricks.

*   I concur with the removal of setting `errno` before `chown`. I checked the other uses
    of clearing `errno` and concur that they need to remain.

LeeTibbert commented on javalib/src/main/scala/java/nio/file/attribute/UserPrincipalNotFoundException.scala:
```diff
@@ -0,0 +1,9 @@
+class UserPrincipalNotFoundException(name: String) extends IOException {
+  def getName(): String = name
+
+  override def getMessage(): String = s"Lookup for $name failed"
```
This line should be deleted.  SN javalib code attempts to emulate Java 8 JVM as closely as
feasible.  I checked the JVM and its getMessage() does a simple pass through. 
Such an override would be a best-practice in other areas of SN, say posixlib.

avdv replied:
Done.

LeeTibbert commented on javalib/src/main/scala/scala/scalanative/nio/fs/attribute/NativeUserPrincipalLookupService.scala:
```diff
@@ -0,0 +1,108 @@
+import scalanative.posix.sys.stat
+import scala.scalanative.nio.fs.UnixException
+
+final case class NativeUserPrincipal(uid: stat.uid_t)(name: Option[String]) extends UserPrincipal {
```
This class and the one immediately below should have  `private` between the class name and the opening
parenthesis for the arguments.  As they are now, the arguments are visible to the public, which does not
match the Java 8 description.

avdv replied:
Fixed.

LeeTibbert commented on nativelib/src/main/resources/pwd.c:
```diff
@@ -20,17 +21,19 @@ void scalanative_passwd_copy(struct passwd *passwd,
 }
 
 int scalanative_getpwnam(char *name, struct scalanative_passwd *buf) {
+    errno = 0;
```
Unless I am missing something, all three edits in the file should be deleted.
The clearing of errno could be author's perogative

Changing the  non-standard handling in the getpwuid code from two values to three valued and
then never checking anything other than the 0 result seems to add complexity with little benefit.

avdv replied:
Yes, you're right. I just forget to revert this file too besides `grp.c`.

avdv replied:
Fixed.

LeeTibbert commented on unit-tests/src/test/scala/java/nio/file/attribute/UserPrincipalLookupServiceSuite.scala:
```diff
@@ -0,0 +1,27 @@
+      lookupService.lookupPrincipalByName("root").getName == "root")
+  }
+
+  test("lookupPrincipalByName throws exception for `gobbledygook` user") {
```
To be friendly to developers running the test suite on systems other than Travis, the two
tests expected to throw should execute only on Travis.

Adapting the Travis detection  idiom kindly supplied by  @francisdb on SN gitter:
(needs scalafmt)
```
 val isTravis = System.getenv("TRAVIS") == "true"
 
  test("lookupPrincipalByName throws exception for `gobbledygook` user") {
      if (isTravis) {
          assertThrows[UserPrincipalNotFoundException](
          lookupService.lookupPrincipalByName("gobbledygook"))
     }
}
```

avdv replied:
Fixed.

LeeTibbert commented on unit-tests/src/test/scala/java/nio/file/attribute/UserPrincipalLookupServiceSuite.scala:
```diff
@@ -0,0 +1,27 @@
+    assertThrows[UserPrincipalNotFoundException](
+      lookupService.lookupPrincipalByGroupName("gobbledygook"))
+  }
+}
```
Studying the test suite I kept asking myself:  What confidence do I/we have that the problem described
as a reproducer in the original commit is actually both fixed and shown to be fixed?

After a number of design studies, I believe a simple solution is to add a test here which does a 
lookup of a strictly numeric group name believed to not exist on Travis, say  65501 (not 65534).

This test would get a short comment that it shows the reproducer in the base commit note
is fixed.  Yes, future devos will have to spend time figuring out this is true. It is not face evident
but a comment will give some future devo a clue & starting place.

The second edit would be to edit the base commit to say that identify the test & file that gives the
essence of the long form reproducer given.  I would keep that long form raw reproducer.

[needs scalafmt]

test("lookupPrincipalByGroupName succeeds for gid only group 65501") {
 if (isTravis) {
    val expected = "65501"
    assert(
      lookupService.lookupPrincipalByGroupName(expected).getName == expected)
  }
}

avdv replied:
To be honest, I thought this will just not work since you cannot simply pass a UID / GID for name.

But you're right. There's a crazy [loophole](https://github.com/keerath/openjdk-8-source/blob/5f6e9d42a9f6b6736100c9c6f43f5f5ea1570cfb/jdk/src/solaris/classes/sun/nio/fs/UnixUserPrincipals.java#L150-L157) (which is not even documented), but you can create a User or Group principal if you pass an arbitrary integer value as String to the `lookupByName` or `lookupByGroupName` methods of the `UserPrincipalLookupService`. :scream:

avdv replied:
I think, in this light, we should add tests which check that for any Integer.toString argument, both functions do not throw and return a user with the given "name".

Alas, that does not really reproduce the initial problem. It doesn't even hit the same code path. The initial problem was to lookup a user/group name for a UID/GID, respectively. For that matter we would need to exercise the `getUsername` and `getGroupName` methods which are only called if the `NativeUserPrincipal`'s `name` is `None`.

Is it possible to access the classes in the `scala.scalanative.nio.fs.attribute` package from the unit-tests?

LeeTibbert replied:
@avdv 
The `Very important notice about Javalib` of the [Contributing Guidelines](https://scala-native.readthedocs.io/en/v0.3.9-docs/contrib/contributing.html#very-important-notice-about-javalibl) say that we can not
use or look at OpenJDk code, so we can not use either the code nor the knowledge of the loophole.

However, we can use knowledge from before looking at the code.  A careful reading of the Java * docs
indicates that all Principals have a name.  The information you provided in the base note indicates that
the name used for gid when there is no other in /etc/groups is gid.

I checked the discontinued Apache Harmony javalib implementation, which we _can_ look at. Its Principals
all throw if a name is not provide.

I think the concept you have surfaced that all Principals have a name allows a major simplification of the code,

LeeTibbert replied:
yes, you are correct, with this generation of changes, the test to prove the base issue was fixed I proposed
is no longer useful.  I'll try again.

LeeTibbert commented:
@avdv  This PR is almost ready to go!  I hope that you see value in the review comments and are not
discouraged.

avdv commented:
> @avdv This PR is almost ready to go! I hope that you see value in the review comments and are not
> discouraged.

No worries! Your insights are very much appreciated. I just need to find some time to spare to go over this PR again. Probably next weekend.

LeeTibbert commented:
@avdv  Thank you for the changes.   

If you are willing to give this another evolution, can we chat?  Either  a private direct message
on [SN gitter](https://gitter.im/scala-native/scala-native)?    

I think a massive simplification is possible with a few edits. and would like to discuss it with you if you have not run out of time & patience for this PR.  

At this point, I think that the only thing missing is a unit test showing that the underlying 
issue is solved.    I'll start studying a good way to do that test, I already know of one bad way ;-)

LeeTibbert commented:
@avdv  

Thank you for all your hard work and, especially, your patience & energy going
through the review process.

This "Looks Good To Me"  LGTM.

* The code is clear, efficient, & maintainable.

* Documentation has been addressed. The commit message at the top of this PR
  contains a clear reproducer. The introduced exception is now documented in
   javalib.rst.

* Two unit-test cases show that this PR address the original reported issue and
   its logical kin.  

@densh 

1)  The two Travis CI failures look like they are Travis issues and have nothing to due with
      this PR, except that it was the one executing. 

2) The two new unit-test cases use `sudo chown` to change the ownership of a newly
     created file.  They are active in this PR as-is, to show that the reported problem & its
     logical sibling are fixed.   

    The `sudo` is Travis-only to prevent execution in J-Random environments, possibly
    with false failures.

    If having two Travis unit-test cases do `sudo` those two tests can be commented out
    and left in-inplace as dormant executable documentation, similar to data generator
    comment blocks at the bottom of files.   A judgement call. 

3) I know that you are fulfilling the demands & requirements of Real Life. 
    When the time comes around for SN again, please let us know if there is anything either 
    @avdv or I need to do.  Thank you (and every success in RL).

avdv commented:
Hi.

I have just rebased against master and cleaned up the PR branch a bit: removed the back-and-forth of code changes and formatted each commit with scalafmt. There should be no functional changes (if I made no mistake); however, the original branch is still available [here](https://github.com/avdv/scala-native/tree/avoid-ioexception_v1).

LeeTibbert commented:
I believe that this Issue is fixed by PR #2244 and can be closed
